fairlearn/postprocessing/_threshold_optimizer.py

```diff
@@ -427,6 +260,167 @@ def _threshold_optimization_equalized_odds(sensitive_features, labels, scores, g
     return predicted_EO_by_sensitive_feature
 
 
+_SUPPORTED_CONSTRAINTS = {
+    DEMOGRAPHIC_PARITY: _threshold_optimization_demographic_parity,
+    EQUALIZED_ODDS: _threshold_optimization_equalized_odds}
+
+
+def _get_soft_predictions(estimator, X):
+    """Returns soft predictions of a classifier using either `predict_proba`
+    or `decision_function` methods.
+    """
+    if hasattr(estimator, "predict_proba"):
+        return estimator.predict_proba(X)
+    elif hasattr(estimator, "decision_function"):
+        return estimator.decision_function(X)
+    else:
+        raise ValueError("{} provides neither a `predict_proba` nor a "
+                            "`decision_function` method.".format(
+                                estimator.__class__.__name__
+                            ))
+
+
+class ThresholdOptimizer(ClassifierMixin, BaseEstimator):
+    """A classifier based on the threshold optimization approach.
+
+    The procedure followed is described in detail in
+    `Hardt et al. (2016) <https://arxiv.org/abs/1610.02413>`_.
+
+    :param unconstrained_predictor: The trained predictor whose output will be
+    post processed
+    :type unconstrained_predictor: A trained predictor
+
+    :param estimator: An untrained estimator that will be trained, and
+        subsequently its output will be post processed
+    :type estimator: An untrained estimator
+
+    :param grid_size: The number of ticks on the grid over which we evaluate the
+        curves. A large grid_size means that we approximate the actual curve, so
+        it increases the chance of being very close to the actual best solution.
+    :type grid_size: int
+
+    :param flip: Allow flipping to negative weights if it improves accuracy.
+    :type flip: bool
+
+    :param plot: Show ROC/selection-error plot if True
+    :type plot: bool
+
+    :param random_state: set to a constant for reproducibility
+    :type random_state: int, np.RandomState, or None
+    """
+
+    def __init__(self, *, estimator=None, constraints=DEMOGRAPHIC_PARITY,
+                 grid_size=1000, flip=True, plot=False, random_state=None):
+
```

Comment:
We have lost `unconstrained_predictor` and all of the related logic. Was this intentional? The idea is that if this argument is present, we should use it in `fit` to initialize `self._estimator` (without calling `self.unconstrained_predictor.fit(...)`).

Reply:
@adrinjalali can elaborate but from what I understand scikit-learn doesn't take already fitted models, but rather fits them itself. We can still have another version where we take an already fitted one, but it wouldn't be scikit-learn compatible. Is that about accurate?

Reply:
Yes. `GridSearchCV` _clones_ the estimators before fitting them, and that clears all the state they have. If we want to allow the user to re-fit on using a pre-fitted estimator, we can implement a `warm_start` strategy for the estimator to use the previously fitted one. But by default all of these should be fit from scratch, since grid search is fitting them on a new part of the data each time anyway.

Reply:
That sounds reasonable. Would be sad to lose the existing functionality if we're planning to add it back in anyway... any chance we can add that warm start in this PR, or is that a major change?

Reply:
added `warm_start`

Reply:
Sorry for a delay on this, but I think that `warm_star` is confusing in this context. 

Just to state the obvious: the high-level semantics of `ThresholdOptimizer` (after it is fitted) is really of a two-stage pipeline. The first step is implemented by `self._estimator.predict()`. It takes as input `X` and returns `y`. The returned `y` is then transformed into `y'` (inside `self.predict()`).

We would like to enable two styles of fitting this pipeline: (1) fitting both steps from the provided data; or (2) fitting only the second step from the provided data (and using the provided `unconstrained_predictor` as the first step). Now if I think about a warm start, I would expect re-fitting a previously fitted pipeline, using the previous pipeline as a warm start. This to me is quite different from (2), which is just fitting the second stage of the pipeline (from scratch) while keeping the first stage of the pipeline untouched.

I'm not quite sure I understand the issue with the interaction between the cloning behavior and the `unconstrained_predictor` argument.

Reply:
maybe you're referring to this?
https://scikit-learn.org/stable/glossary.html#term-metaestimators

Reply:
To help resolve this. I suggest that we also implement the pattern from `CalibratedClassifierCV` [link](https://github.com/scikit-learn/scikit-learn/blob/b194674c42d54b26137a456c510c5fdba1ba23e0/sklearn/calibration.py#L32)

That would mean, using `prefit=False` instead of `warm_Start=False`.

Reply:
Sorry took me a while to concentrate enough to understand what you mean.

I understand that `warm_start` is different from what we have here. I'm just not sure why it's really needed. In a usual pipeline, we usually don't have pre-fitted estimators to feed to meta-estimators. Let me open a separate issue to talk about it there.